key.go

```diff
@@ -0,0 +1,810 @@
+		case AlgorithmES384:
+			curve = elliptic.P384()
+		case AlgorithmES512:
+			curve = elliptic.P521()
+		}
+
+		priv := &ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{Curve: curve, X: new(big.Int), Y: new(big.Int)},
+			D:         new(big.Int),
+		}
+		priv.X.SetBytes(k.X)
+		priv.Y.SetBytes(k.Y)
+		priv.D.SetBytes(k.D)
+
+		return priv, nil
+	case AlgorithmEd25519:
+		buf := make([]byte, ed25519.PrivateKeySize)
+
+		copy(buf, k.D)
+		copy(buf[32:], k.X)
+
+		return ed25519.PrivateKey(buf), nil
+	default:
+		return nil, ErrAlgorithmNotSupported
+	}
+}
+
+// GetAlgorithm returns the Algorithm associated with Key. If Key.Algorithm is
+// set, that is what is returned. Otherwise, the algorithm is inferred using
+// Key.Curve. This method does NOT validate that Key.Algorithm, if set, aligns
+// with Key.Curve.
+func (k *Key) GetAlgorithm() (Algorithm, error) {
+	if k.Algorithm != AlgorithmInvalid {
+		return k.Algorithm, nil
+	}
+
+	return k.deriveAlgorithm()
+}
+
+// Signer returns a Signer created using Key.
+func (k *Key) Signer() (Signer, error) {
+	if err := k.Validate(); err != nil {
+		return nil, err
+	}
+
+	if k.KeyOps != nil {
+		signFound := false
+
+		for _, kop := range k.KeyOps {
+			if kop == KeyOpSign {
+				signFound = true
+				break
+			}
+		}
+
+		if !signFound {
+			return nil, ErrSignOpNotSupported
+		}
+	}
+
+	priv, err := k.PrivateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	alg, err := k.GetAlgorithm()
+	if err != nil {
+		return nil, err
+	}
+
+	var signer crypto.Signer
+	var ok bool
+
+	switch alg {
+	case AlgorithmES256, AlgorithmES384, AlgorithmES512:
+		signer, ok = priv.(*ecdsa.PrivateKey)
+		if !ok {
+			return nil, ErrInvalidPrivKey
+		}
+	case AlgorithmEd25519:
+		signer, ok = priv.(ed25519.PrivateKey)
+		if !ok {
+			return nil, ErrInvalidPrivKey
+		}
+	default:
+		return nil, ErrAlgorithmNotSupported
+	}
+
+	return NewSigner(alg, signer)
+}
+
+// Verifier returns a Verifier created using Key.
+func (k *Key) Verifier() (Verifier, error) {
+	if err := k.Validate(); err != nil {
+		return nil, err
+	}
+
+	if k.KeyOps != nil {
+		verifyFound := false
+
+		for _, kop := range k.KeyOps {
+			if kop == KeyOpVerify {
+				verifyFound = true
+				break
+			}
+		}
+
+		if !verifyFound {
+			return nil, ErrVerifyOpNotSupported
+		}
+	}
+
+	pub, err := k.PublicKey()
+	if err != nil {
+		return nil, err
+	}
+
+	alg, err := k.GetAlgorithm()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewVerifier(alg, pub)
+}
+
+// deriveAlgorithm derives the intended algorithm for the key from its curve.
+func (k *Key) deriveAlgorithm() (Algorithm, error) {
+	switch k.KeyType {
+	case KeyTypeEc2, KeyTypeOkp:
+		switch k.Curve {
+		case CurveP256:
+			return AlgorithmES256, nil
+		case CurveP384:
+			return AlgorithmES384, nil
+		case CurveP521:
+			return AlgorithmES512, nil
+		case CurveEd25519:
+			return AlgorithmEd25519, nil
+		default:
+			return AlgorithmInvalid, fmt.Errorf("unsupported curve %q", k.Curve.String())
+		}
```

Comment:
To be rigorous, could you document the reference to this mapping?

[RFC 8152 section 8.1](https://datatracker.ietf.org/doc/html/rfc8152#section-8.1) states that

> In order to promote interoperability, it is suggested that SHA-256 be
>    used only with curve P-256, SHA-384 be used only with curve P-384,
>    and SHA-512 be used with curve P-521.  This is aligned with the
>    recommendation in [Section 4 of [RFC5480]](https://datatracker.ietf.org/doc/html/rfc5480#section-4).

In other words, the curve `P-521` can be combined with `SHA-256` as `ES256` although it is not suggested... but possible.